specs/switch-driver-capabilities.md

```diff
@@ -0,0 +1,48 @@
+
+# Expose network driver capabilites
+-----------
+
+[Issue 410](https://github.com/CCI-MOC/hil/issues/410)
+
+We want to expose switch driver capabilities. Capabilities include:
+
+- Jumbo frames are per-switch on our Dell switches and per-port on our Ciscos (#755)
+-  QoS (#384)
+- Disabling STP (#779)
+- DHCP/ARP Snooping (#471)
+- Enabling the tagging of DHCP packets with the port number (from @okrieg or @pjd-nu)
+
+# Solution
+-----------
+
+* Every switch driver will have a list called `capabilities` in the class. So it would look like
+`capabilities = ["require-native-network", "per-port-jumbo-frames"]`
+
+* and then there'll be a method  `get_capabilities()` that returns a boolean if a capability exists or not by checking it in the list.
+
+* an API that needs to test this, can call that method `switch.get_capabilities('require-native-network`) to ensure that the switch supports that method, and then proceed with the original request or error out.
+
+* `show_node()` can be updated to show the capabilities they support.
+
+# Alternative Solutions
+-----------------------
+
+@henn proposed that we store the capabilites in a database and those will be mapped
+to switch objects. That way we can control indiviual switches instead of a whole
+class of switches.
+
+Rest of the steps would remain the same.
+
+# Arch Impact
+-----------------
+
+Are there any new architectural assumptions we are now making?
+I am making an assumption that we would treat all switches that belong to a class
+the same. That is, all brocade switches would have those capabilities as specified
+in the Class.
```

Comment:
If we don't make the list itself public, and only specify the `have_capability` method, drivers could report the capability based on other information stored about the switch, so they wouldn't necessarily have to return the same value for all switches. That makes it unclear how we would list the caps in `show_node` though. Maybe we can just have the method `get_capabilities`, that returns the list, but not actually make the list itself public?

Reply:
> drivers could report the capability based on other information stored about the switch

where would this other information be stored? in the switch table?

> Maybe we can just have the method get_capabilities, that returns the list, but not actually make the list itself public?

yep, and that will be called by `show_node()`

Reply:
Yeah, probably in a driver-specific table.

Reply:
Hmm, in that case we would need multiple tables. A table to store switch info (we already have that), a table to store all possible switch capabilities, and then a table to map capabilities to switches.

``` SQL
CREATE TABLE Switches(SwitchID integer PRIMARY KEY, SwitchName text);

/* Create few records in this table */
INSERT INTO Switches VALUES(1,'Brocade');
INSERT INTO Switches VALUES(2,'Dell');

CREATE TABLE AllSwitchCapabilities(CapID integer PRIMARY KEY, Capability text);

INSERT INTO AllSwitchCapabilities Values(1,'jumbo');
INSERT INTO AllSwitchCapabilities Values(2,'native');
INSERT INTO AllSwitchCapabilities Values(3,'something');

CREATE TABLE SwitchCapability(ID integer PRIMARY KEY, SwitchID int, CapID int);
INSERT INTO SwitchCapability Values(1, 2, 1);
INSERT INTO SwitchCapability Values(2, 2, 2);
```
and then the query to list all capabilities would look like this:

```SQL
SELECT SwitchName, Capability from Switches, AllSwitchCapabilities, SwitchCapability where
Switches.SwitchID == SwitchCapability.SwitchID and SwitchCapability.CapID == 
AllSwitchCapabilities.CapID

```

Is there a simpler way to do this?

Reply:
Depending on the switch, you could just have a column of type text that stores the list encoded as JSON. The point is that implementations can do whatever makes sense for them.

Reply:
I think you misunderstood what I was getting at. The notion is that get_capabilities would be implemented by the switch driver. Nothing needs to be added to the Switch table; if schema changes occur they would be isolated to the driver. But the point is it's an implementation detail; by making it a function the driver can do whatever it wants to create that list. The driver *could* store the info in the database (under one of its own tables), or it could just return a hard-coded list (probably makes sense for most drivers), or compute it some other way.

Reply:
> The notion is that get_capabilities would be implemented by the switch driver.

Yes, I agree. And like you said, it could get/generate this list from a database, or a hardcoded list in the Class.

What I am trying to address is that, let's say I have 2 switches of type brocade, and for whatever reason I want to enable a certain capability only on one of them. 

> or it could just return a hard-coded list (probably makes sense for most drivers),

If I hard-code this list in the switch driver, then I can't differentiate between the capabilities of 2 different switches of the same type. So I must store it somewhere, right? And I thought keeping it as JSON in the db made sense.

Or should we not expose the per-switch capability in HIL? and enforce same capabilities if they belong to the same Class.

Reply:
Just catching up on this thread.

I agree with @zenhack - having driver API calls gives us a lot of flexibility in how we keep the backend. Please see my comments in [the code](https://github.com/CCI-MOC/hil/pull/876#discussion_r140098278). Also, by giving it an API, the switch driver can do whatever it wants. For now, we can keep a hard-coded list in the class, and in the future we can use a per-driver or global DB table to make it per-switch instantiation.